crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,222 @@
-        let encoding = snapshot.encoding();
-        let version = document.version();
-        let actions: Result<Vec<_>> = params
-            .context
-            .diagnostics
-            .into_iter()
-            .map(|diagnostic| {
-                let Some(data) = diagnostic.data else {
-                    return Ok(None);
-                };
-                let diagnostic_fix: crate::lint::DiagnosticFix = serde_json::from_value(data)
-                    .map_err(|err| anyhow::anyhow!("failed to deserialize diagnostic data: {err}"))
-                    .with_failure_code(lsp_server::ErrorCode::ParseError)?;
-                let edits = diagnostic_fix
-                    .fix
-                    .edits()
-                    .iter()
-                    .map(|edit| types::TextEdit {
-                        range: edit.range().to_range(
-                            document.contents(),
-                            document.index(),
-                            encoding,
-                        ),
-                        new_text: edit.content().unwrap_or_default().to_string(),
-                    });
-
-                let changes = vec![types::TextDocumentEdit {
-                    text_document: types::OptionalVersionedTextDocumentIdentifier::new(
-                        url.clone(),
-                        version,
-                    ),
-                    edits: edits.map(types::OneOf::Left).collect(),
-                }];
-
-                let title = diagnostic_fix
-                    .kind
-                    .suggestion
-                    .unwrap_or(diagnostic_fix.kind.name);
-                Ok(Some(types::CodeAction {
-                    title,
-                    kind: Some(types::CodeActionKind::QUICKFIX),
+        // compute the associated document edits for each diagnostic
+        // these will get re-used when building the actual code actions afterwards
+        let edits = diagnostic_edits(
+            document,
+            snapshot.url(),
+            snapshot.encoding(),
+            document.version(),
+            params.context.diagnostics.into_iter(),
+        )
+        .collect::<Result<Vec<_>>>()?;
+
+        let mut response: types::CodeActionResponse = types::CodeActionResponse::default();
+
+        for action in available_actions {
+            match action {
+                SupportedCodeAction::QuickFix => response.extend(quick_fix(edits.iter())),
+                SupportedCodeAction::FixAll => response.extend(fix_all(edits.iter())),
+                SupportedCodeAction::OrganizeImports => {
+                    todo!("Implement the `source.organizeImports` code action")
+                }
+            }
+        }
+
+        Ok(Some(response))
+    }
+}
+
+fn diagnostic_edits<'d>(
+    document: &'d DocumentRef,
+    url: &'d Url,
+    encoding: PositionEncoding,
+    version: DocumentVersion,
+    diagnostics: impl Iterator<Item = types::Diagnostic> + 'd,
```

Comment:
Nit: I think I would just go with a `vec` here to keep things simple. We don't have any usages today where we want to pass an iterator
```suggestion
    diagnostics: Vec<types::Diagnostic>,
```